lib/auth/auth_with_roles.go

```diff
@@ -661,6 +662,45 @@ func (a *ServerWithRoles) GetNodes(ctx context.Context, namespace string, opts .
 	return filteredNodes, nil
 }
 
+// ListNodes returns a paginated list of nodes filtered by user access.
+func (a *ServerWithRoles) ListNodes(ctx context.Context, namespace string, limit int, startKey string) (page []types.Server, nextKey string, err error) {
+	if err := a.action(namespace, types.KindNode, types.VerbList); err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+
+	// Retrieve and filter pages of nodes until we can fill a page or run out of nodes.
+	page = make([]types.Server, 0, limit)
+	for len(page) != limit {
+		nextPage, nextKey, err := a.authServer.ListNodes(ctx, namespace, limit, startKey)
```

Comment:
Not: The fact that the auth server may unnecessarily load all nodes many times for the calculation of one page here feels problematic.  I think it's fine for now, but let's add a TODO to move toward an iterator abstraction for `Get/ListNodes` in the future.  That's out of scope for this patch, but would probably be good for all kinds of reasons, including abstracting over truly paginated data, and artificially paginated slices.